SSL: optional ssl_client_certificate for ssl_verify_client.

praveen-li opened the pull request:
Starting from TLSv1.1 (as seen since draft-ietf-tls-rfc2246-bis-00),
the "certificate_authorities" field grammar of the CertificateRequest
message was redid to allow no distinguished names.  In TLSv1.3, with
the restructured CertificateRequest message, this can be similarly
done by optionally including the "certificate_authorities" extension.
This allows to avoid sending DNs at all.

In practice, aside from published TLS specifications, all supported
SSL/TLS libraries allow to request client certificates with an empty
DN list for any protocol version.  For instance, when operating in
TLSv1, this results in sending the "certificate_authorities" list as
a zero-length vector, which corresponds to the TLSv1.1 specification.
Such behaviour goes back to SSLeay.

The change relaxes the requirement to specify at least one trusted CA
certificate in the ssl_client_certificate directive, which resulted in
sending DNs of these certificates (closes #142).  Instead, all trusted
CA certificates can be specified now using the ssl_trusted_certificate
directive if needed.  A notable difference that certificates specified
in ssl_trusted_certificate are always loaded remains (see [3648ba](https://github.com/nginx/nginx/commit/3648ba7db833d318269daba2a8d6be42660c5b60).)


### Proposed changes

As per below RFCs, ssl_client_certificate should be optional with TLS1.1+ when ssl_verify_client is configured as on/optional:

TLS1.3: rfc8446#section-4.2.4:
```
 The server MAY send it(CA) in the  CertificateRequest message.
```

TLS1.2: rfc5246#section-7.4.4:
```
If the certificate_authorities list is empty, then the client MAY send any certificate of the appropriate ClientCertificateType, unless there is some external arrangement to the contrary.
```
TLS1.1: rfc4346#section-7.4.4:  Similar wording as TLS 1.2.

TLSv1.0: rfc2246#section-7.4.4: It was not clearly defined in v1.0, that CA list in certificate request message can be empty or not.

Also Nginx documentation seems to suggest the same:
```
The list of certificates will be sent to clients. If this is not desired, the [ssl_trusted_certificate] directive can be used."
```




**Tested with nginx-tests.
--------------------------
Verified after deploying on Debian.**
--------------------------

Relevant Config
```
       # Enable client certificate verification (mTLS)
        ssl_trusted_certificate /Users/pchere/Desktop/src_code/certs/ca.crt;  # CA cert for client certificates
        ssl_verify_client on;  # Force client certificate validation
```

With ssl_protocols TLSv1.2 TLSv1.3;
```
% sudo /usr/local/nginx/sbin/nginx -s reload

%curl -v --cert /Users/pchere/Desktop/src_code/certs/client.crt --key /Users/pchere/Desktop/src_code/certs/client.key --cacert /Users/pchere/Desktop/src_code/certs/ca.crt https://localhost/
* (304) (OUT), TLS handshake, Client hello (1):
*  CAfile: /Users/pchere/Desktop/src_code/certs/ca.crt
*  CApath: none
* (304) (IN), TLS handshake, Server hello (2):
* (304) (IN), TLS handshake, Unknown (8):
* (304) (IN), TLS handshake, Request CERT (13):
* (304) (IN), TLS handshake, Certificate (11):
* (304) (IN), TLS handshake, CERT verify (15):
* (304) (IN), TLS handshake, Finished (20):
* (304) (OUT), TLS handshake, Certificate (11):
* (304) (OUT), TLS handshake, CERT verify (15):
* (304) (OUT), TLS handshake, Finished (20):
* SSL connection using TLSv1.3 / AEAD-AES256-GCM-SHA384 / [blank] / UNDEF
* ALPN: server accepted http/1.1
```


Comment ssl_trusted_certificate in nginx config.
```
 % sudo /usr/local/nginx/sbin/nginx -s reload
nginx: [emerg] no ssl_client_certificate or ssl_trusted_certificatefor ssl_verify_client
```

praveen-li commented on src/http/modules/ngx_http_ssl_module.c:
```diff
@@ -788,9 +788,18 @@ ngx_http_ssl_merge_srv_conf(ngx_conf_t *cf, void *parent, void *child)
-            ngx_log_error(NGX_LOG_EMERG, cf->log, 0,
-                          "no ssl_client_certificate for ssl_verify_client");
-            return NGX_CONF_ERROR;
+            if (conf->protocols & (NGX_SSL_SSLv3|NGX_SSL_TLSv1))
```
Avoided check for SSLv2 and added for SSLv3 TLSv1, since mTLS is not supported with SSLv2. 
Considering if  "ssl_protocols SSLv2 TLSv1.3;" is configured, then config should react as per TLS1.3 for ssl_client_certificate.

praveen-li commented:
@pluknet Kindly review this PR.

Many deployments may have dependency on a.) this fix and b.) client certificate validation with OCSP. Kindly let me know, if you recommend breaking this into http, mail and stream.

Thanks a ton in advance.

praveen-li commented:
@arut  

Hi Roman, It will be really helpful, if you can review this PR. We have a product, where we do not want to force user to configure the upper level config which maps to ssl_client_certificate in nginx for mTLS. 

Considering, from TLS1.1+ (rfc4346#section-7.4.4), it is not mandatory to send 'CA Issuer names' in 'certificate request' message from server to client in case on mTLS. So, It will be great if we can make ssl_client_certificate directive optional when ssl_trusted_certificate is configured. 

Thanks in advance for reviewing it.

pluknet commented on src/http/modules/ngx_http_ssl_module.c:
```diff
@@ -788,9 +788,18 @@ ngx_http_ssl_merge_srv_conf(ngx_conf_t *cf, void *parent, void *child)
+                            "no ssl_client_certificate or "
+                            "ssl_trusted_certificatefor ssl_verify_client");
+                return NGX_CONF_ERROR;
+            }
```
I don't think it makes sense to separate logic for older protocol versions, because this complicates code for no apparent reason:
- legacy setups will barely change using legacy configs, it should be safe to expand to all protocol versions, even if it might not work
- empty DN list works in practice with older protocol versions, SSL/TLS library handles that for us.
- if any legacy clients exist that cannot accept TLSv1.1 AND empty DN lists, then see the 1st point about legacy setups. 
As such, we can fold a single config check.
```suggestion
        if (conf->verify != 3
            && conf->client_certificate.len == 0
            && conf->trusted_certificate.len == 0)
        {
            ngx_log_error(NGX_LOG_EMERG, cf->log, 0,
                          "no ssl_client_certificate or "
                          "ssl_trusted_certificate for ssl_verify_client");
            return NGX_CONF_ERROR;
        }
```

praveen-li replied:
@pluknet Thanks a lot for finding time to review it. I express same concern in [this thread.](https://mailman.nginx.org/pipermail/nginx-devel/2024-August/6VF2IN3A24UZKVLQ3IUUL7AJY6L5QS2Z.html). 

Where I asked below. So, I already agree with you.
```
Also, let me know if we need to keep
check simple
as this:
*if (conf->client_certificate.len == 0 && conf->trusted_certificate.len ==
0 && conf->verify != 3) {*
```

praveen-li replied:
@pluknet I will reraise-PR with suggested commit today.  

AFAIK, as per contributing guide, in this case, it has to be a forced push to make it one commit. Kindly let me know, if we prefer "Squash and Merge" instead, which can have 2 commits, i.e. 2nd commit to address review comments.

pluknet replied:
Generally we ask to break changes in logically-separate self-sufficient commits to maintain a clean history.
As per CONTRIBUTING.md:
```
- Keep a clean, concise and meaningful commit history on your branch, rebasing
locally and breaking changes logically into commits before submitting a PR
```
So, you were correct to make a force push with a single commit.

pluknet commented on src/http/modules/ngx_http_ssl_module.c:
```diff
@@ -788,9 +788,18 @@ ngx_http_ssl_merge_srv_conf(ngx_conf_t *cf, void *parent, void *child)
-            ngx_log_error(NGX_LOG_EMERG, cf->log, 0,
-                          "no ssl_client_certificate for ssl_verify_client");
-            return NGX_CONF_ERROR;
+            if (conf->protocols & (NGX_SSL_SSLv3|NGX_SSL_TLSv1))
```
style: opening curly brace should be on the same line (unless wrapped)

praveen-li replied:
I will address it in suggested commit and will follow the same in future. Thanks.

pluknet commented on src/http/modules/ngx_http_ssl_module.c:
```diff
@@ -788,9 +788,18 @@ ngx_http_ssl_merge_srv_conf(ngx_conf_t *cf, void *parent, void *child)
+            if (conf->protocols & (NGX_SSL_SSLv3|NGX_SSL_TLSv1))
+            {
+                ngx_log_error(NGX_LOG_EMERG, cf->log, 0,
+                            "no ssl_client_certificate for ssl_verify_client");
```
style: continued arguments should indent under the first argument

praveen-li replied:
I will address it in suggested commit and will follow the same in future. Thanks.

pluknet commented on src/http/modules/ngx_http_ssl_module.c:
```diff
@@ -788,9 +788,18 @@ ngx_http_ssl_merge_srv_conf(ngx_conf_t *cf, void *parent, void *child)
+                            "no ssl_client_certificate for ssl_verify_client");
+                return NGX_CONF_ERROR;
+            }
+            if (conf->trusted_certificate.len == 0) {
```
style: empty line between code blocks needed

praveen-li replied:
I will address it in suggested commit and will follow the same in future. Thanks.

pluknet commented:
Suggested commit:
```
changeset:   143ac35cb (HEAD -> make_ssl_client_certificate_optional, sandbox/master)
user:        Sergey Kandaurov <pluknet@nginx.com>
date:        Wed Sep 18 18:30:09 2024 +0400
description:
SSL: optional ssl_client_certificate for ssl_verify_client.

Starting from TLSv1.1 (as seen since draft-ietf-tls-rfc2246-bis-00),
the "certificate_authorities" field grammar of the CertificateRequest
message was redid to allow no distinguished names.  In TLSv1.3, with
the restructured CertificateRequest message, this can be similarly
done by optionally including the "certificate_authorities" extension.
This allows to avoid sending DNs at all.

In practice, aside from published TLS specifications, all supported
SSL/TLS libraries allow to request client certificates with an empty
DN list for any protocol version.  For instance, when operating in
TLSv1, this results in sending the "certificate_authorities" list as
a zero-length vector, which corresponds to the TLSv1.1 specification.
Such behaviour goes back to SSLeay.

The change relaxes the requirement to specify at least one trusted CA
certificate in the ssl_client_certificate directive, which resulted in
sending DNs of these certificates (closes #142).  Instead, all trusted
CA certificates can be specified now using the ssl_trusted_certificate
directive if needed.  A notable difference that certificates specified
in ssl_trusted_certificate are always loaded remains (see 3648ba7db).


diff --git a/src/http/modules/ngx_http_ssl_module.c b/src/http/modules/ngx_http_ssl_module.c
index 1c92d9fa8..abc8d49ab 100644
--- a/src/http/modules/ngx_http_ssl_module.c
+++ b/src/http/modules/ngx_http_ssl_module.c
@@ -787,9 +787,13 @@ ngx_http_ssl_merge_srv_conf(ngx_conf_t *cf, void *parent, void *child)
 
     if (conf->verify) {
 
-        if (conf->client_certificate.len == 0 && conf->verify != 3) {
+        if (conf->verify != 3
+            && conf->client_certificate.len == 0
+            && conf->trusted_certificate.len == 0)
+        {
             ngx_log_error(NGX_LOG_EMERG, cf->log, 0,
-                          "no ssl_client_certificate for ssl_verify_client");
+                          "no ssl_client_certificate or "
+                          "ssl_trusted_certificate for ssl_verify_client");
             return NGX_CONF_ERROR;
         }
 
diff --git a/src/mail/ngx_mail_ssl_module.c b/src/mail/ngx_mail_ssl_module.c
index aebb4ccb6..b547dc101 100644
--- a/src/mail/ngx_mail_ssl_module.c
+++ b/src/mail/ngx_mail_ssl_module.c
@@ -450,9 +450,13 @@ ngx_mail_ssl_merge_conf(ngx_conf_t *cf, void *parent, void *child)
 
     if (conf->verify) {
 
-        if (conf->client_certificate.len == 0 && conf->verify != 3) {
+        if (conf->verify != 3
+            && conf->client_certificate.len == 0
+            && conf->trusted_certificate.len == 0)
+        {
             ngx_log_error(NGX_LOG_EMERG, cf->log, 0,
-                          "no ssl_client_certificate for ssl_verify_client");
+                          "no ssl_client_certificate or "
+                          "ssl_trusted_certificate for ssl_verify_client");
             return NGX_CONF_ERROR;
         }
 
diff --git a/src/stream/ngx_stream_ssl_module.c b/src/stream/ngx_stream_ssl_module.c
index 072e74917..0233a9258 100644
--- a/src/stream/ngx_stream_ssl_module.c
+++ b/src/stream/ngx_stream_ssl_module.c
@@ -1008,9 +1008,13 @@ ngx_stream_ssl_merge_srv_conf(ngx_conf_t *cf, void *parent, void *child)
 
     if (conf->verify) {
 
-        if (conf->client_certificate.len == 0 && conf->verify != 3) {
+        if (conf->verify != 3
+            && conf->client_certificate.len == 0
+            && conf->trusted_certificate.len == 0)
+        {
             ngx_log_error(NGX_LOG_EMERG, cf->log, 0,
-                          "no ssl_client_certificate for ssl_verify_client");
+                          "no ssl_client_certificate or "
+                          "ssl_trusted_certificate for ssl_verify_client");
             return NGX_CONF_ERROR;
         }
 
```

praveen-li commented:
> Suggested commit:
> 
> ```
> changeset:   143ac35cb (HEAD -> make_ssl_client_certificate_optional, sandbox/master)
> user:        Sergey Kandaurov <pluknet@nginx.com>
> date:        Wed Sep 18 18:30:09 2024 +0400
> description:
> SSL: optional ssl_client_certificate for ssl_verify_client.
> 
> Starting from TLSv1.1 (as seen since draft-ietf-tls-rfc2246-bis-00),
> the "certificate_authorities" field grammar of the CertificateRequest
> message was redid to allow no distinguished names.  In TLSv1.3, with
> the restructured CertificateRequest message, this can be similarly
> done by optionally including the "certificate_authorities" extension.
> This allows to avoid sending DNs at all.
> 
> In practice, aside from published TLS specifications, all supported
> SSL/TLS libraries allow to request client certificates with an empty
> DN list for any protocol version.  For instance, when operating in
> TLSv1, this results in sending the "certificate_authorities" list as
> a zero-length vector, which corresponds to the TLSv1.1 specification.
> Such behaviour goes back to SSLeay.
> 
> The change relaxes the requirement to specify at least one trusted CA
> certificate in the ssl_client_certificate directive, which resulted in
> sending DNs of these certificates (closes #142).  Instead, all trusted
> CA certificates can be specified now using the ssl_trusted_certificate
> directive if needed.  A notable difference that certificates specified
> in ssl_trusted_certificate are always loaded remains (see 3648ba7db).
> 
> 
> diff --git a/src/http/modules/ngx_http_ssl_module.c b/src/http/modules/ngx_http_ssl_module.c
> index 1c92d9fa8..abc8d49ab 100644
> --- a/src/http/modules/ngx_http_ssl_module.c
> +++ b/src/http/modules/ngx_http_ssl_module.c
> @@ -787,9 +787,13 @@ ngx_http_ssl_merge_srv_conf(ngx_conf_t *cf, void *parent, void *child)
>  
>      if (conf->verify) {
>  
> -        if (conf->client_certificate.len == 0 && conf->verify != 3) {
> +        if (conf->verify != 3
> +            && conf->client_certificate.len == 0
> +            && conf->trusted_certificate.len == 0)
> +        {
>              ngx_log_error(NGX_LOG_EMERG, cf->log, 0,
> -                          "no ssl_client_certificate for ssl_verify_client");
> +                          "no ssl_client_certificate or "
> +                          "ssl_trusted_certificate for ssl_verify_client");
>              return NGX_CONF_ERROR;
>          }
>  
> diff --git a/src/mail/ngx_mail_ssl_module.c b/src/mail/ngx_mail_ssl_module.c
> index aebb4ccb6..b547dc101 100644
> --- a/src/mail/ngx_mail_ssl_module.c
> +++ b/src/mail/ngx_mail_ssl_module.c
> @@ -450,9 +450,13 @@ ngx_mail_ssl_merge_conf(ngx_conf_t *cf, void *parent, void *child)
>  
>      if (conf->verify) {
>  
> -        if (conf->client_certificate.len == 0 && conf->verify != 3) {
> +        if (conf->verify != 3
> +            && conf->client_certificate.len == 0
> +            && conf->trusted_certificate.len == 0)
> +        {
>              ngx_log_error(NGX_LOG_EMERG, cf->log, 0,
> -                          "no ssl_client_certificate for ssl_verify_client");
> +                          "no ssl_client_certificate or "
> +                          "ssl_trusted_certificate for ssl_verify_client");
>              return NGX_CONF_ERROR;
>          }
>  
> diff --git a/src/stream/ngx_stream_ssl_module.c b/src/stream/ngx_stream_ssl_module.c
> index 072e74917..0233a9258 100644
> --- a/src/stream/ngx_stream_ssl_module.c
> +++ b/src/stream/ngx_stream_ssl_module.c
> @@ -1008,9 +1008,13 @@ ngx_stream_ssl_merge_srv_conf(ngx_conf_t *cf, void *parent, void *child)
>  
>      if (conf->verify) {
>  
> -        if (conf->client_certificate.len == 0 && conf->verify != 3) {
> +        if (conf->verify != 3
> +            && conf->client_certificate.len == 0
> +            && conf->trusted_certificate.len == 0)
> +        {
>              ngx_log_error(NGX_LOG_EMERG, cf->log, 0,
> -                          "no ssl_client_certificate for ssl_verify_client");
> +                          "no ssl_client_certificate or "
> +                          "ssl_trusted_certificate for ssl_verify_client");
>              return NGX_CONF_ERROR;
>          }
>  
> ```

@pluknet Agree with the suggested commit as [mentioned in thread](https://mailman.nginx.org/pipermail/nginx-devel/2024-August/6VF2IN3A24UZKVLQ3IUUL7AJY6L5QS2Z.html). Pushed the same commit. Kindly review and merge.  Thanks again.

pluknet commented:
Over to @arut for the second review.

pluknet committed:
SSL: optional ssl_client_certificate for ssl_verify_client.

Starting from TLSv1.1 (as seen since draft-ietf-tls-rfc2246-bis-00),
the "certificate_authorities" field grammar of the CertificateRequest
message was redone to allow no distinguished names.  In TLSv1.3, with
the restructured CertificateRequest message, this can be similarly
done by optionally including the "certificate_authorities" extension.
This allows to avoid sending DNs at all.

In practice, aside from published TLS specifications, all supported
SSL/TLS libraries allow to request client certificates with an empty
DN list for any protocol version.  For instance, when operating in
TLSv1, this results in sending the "certificate_authorities" list as
a zero-length vector, which corresponds to the TLSv1.1 specification.
Such behaviour goes back to SSLeay.

The change relaxes the requirement to specify at least one trusted CA
certificate in the ssl_client_certificate directive, which resulted in
sending DNs of these certificates (closes #142).  Instead, all trusted
CA certificates can be specified now using the ssl_trusted_certificate
directive if needed.  A notable difference that certificates specified
in ssl_trusted_certificate are always loaded remains (see 3648ba7db).

Co-authored-by: Praveen Chaudhary <praveenc@nvidia.com>

arut reviewed (changes requested):
redid -> redone in the commit log.
Otherwise looks ok.

pluknet commented:
Addressed review comments.
